resolver/resolver.cc

```diff
@@ -1144,25 +1144,46 @@ class ResolveConstantsWalk {
         ENFORCE(block->body);
 
         auto blockLoc = core::Loc(todo.file, block->body.loc());
-        auto *id = ast::cast_tree<ast::ConstantLit>(block->body);
+        core::ClassOrModuleRef symbol = core::Symbols::StubModule();
 
-        if (id == nullptr || !id->symbol.exists() || !id->symbol.isClassOrModule()) {
+        if(auto *constant = ast::cast_tree<ast::ConstantLit>(block->body)) {
+            if(constant != nullptr && constant->symbol.exists() && constant->symbol.isClassOrModule()) {
+                symbol = constant->symbol.asClassOrModuleRef();
+            }
+        } else if(ast::MK::isTClassOf(block->body)) {
+            send = ast::cast_tree<ast::Send>(block->body);
+
+            ENFORCE(send);
+
+            if(send->numPosArgs() == 1) {
+                auto arg = send->getPosArg(0).deepCopy();
+                if(auto *argClass = ast::cast_tree<ast::ConstantLit>(arg)) {
+                    if(argClass != nullptr && argClass->symbol.exists() && argClass->symbol.isClassOrModule()) {
+                        if constexpr (isMutableStateType) {
+                            symbol = argClass->symbol.asClassOrModuleRef().data(gs)->singletonClass(gs);
+                        }
+                    }
+                }
+            }
+        }
+
+        if (symbol == core::Symbols::StubModule()) {
             if (auto e = gs.beginError(blockLoc, core::errors::Resolver::InvalidRequiredAncestor)) {
                 e.setHeader("Argument to `{}` must be statically resolvable to a class or a module",
                             send->fun.show(gs));
             }
             return;
         }
 
-        if (id->symbol == owner) {
+        if (symbol == owner) {
             if (auto e = gs.beginError(blockLoc, core::errors::Resolver::InvalidRequiredAncestor)) {
                 e.setHeader("Must not pass yourself to `{}`", send->fun.show(gs));
             }
             return;
         }
```

Comment:
I think we need tests to ensure these errors also fire appropriately for `T.class_of(MODULE)`, e.g. you shouldn't be able to say `requires_ancestor { T.class_of(NonExistentModule) }` nor `requires_ancestor { T.class_of(MyClass) }` inside `MyClass`.

The error messages should probably be slightly different for bare `MODULE` vs. `T.class_of(MODULE)`.

It looks like `test/testdata/resolver/requires_ancestor_errors.rb` has tests like this, so that'd be the appropriate place to add them.